src/libopensc/card-fineid.c

```diff
@@ -0,0 +1,997 @@
+		req[0] = FINEID_HASHING_BY_CARD;
+	else
+		req[0] = FINEID_HASHING_EXTERNALLY;
+
+	req[1] = ilen-ii;
+
+	sc_log(card->ctx, "Finalizing at offset %lu", ii);
+
+	/* Security operation hash
+	 *   INS 2A sec op
+	 *   P1  90 hash
+	 *   P2  A0 last data block
+	 */
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x2A, 0x90, 0xA0);
+	apdu.datalen = reqlen;
+	apdu.data = req;
+	apdu.lc = reqlen;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+	rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(card->ctx, rv, "Last block send failed");
+
+	/* Security operation sign
+	 *   INS 2A sec op
+	 *   P1  9E sign, signature is returned in response
+	 *   P2  9A
+	 */
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0x2A, 0x9E, 0x9A);
+	apdu.le = olen > 256 ? 256 : olen;
+	apdu.resp = resp;
+	apdu.resplen = olen;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+	rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	LOG_TEST_RET(card->ctx, rv, "Signature receiving failed");
+
+	if (apdu.resplen > olen) {
+		sc_log(card->ctx,
+		       "Compute signature failed: invalid response length %"SC_FORMAT_LEN_SIZE_T"u",
+		       apdu.resplen);
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_CARD_CMD_FAILED);
+	}
+
+	memcpy(out, apdu.resp, apdu.resplen);
+
+	LOG_FUNC_RETURN(card->ctx, apdu.resplen);
+}
+
+
+/* Return the default AAK for this type of card */
+static int
+fineid_get_default_key(struct sc_card *card, struct sc_cardctl_default_key *data)
+{
+	LOG_FUNC_RETURN(card->ctx, SC_ERROR_NO_DEFAULT_KEY);
+}
+
+
+static int
+fineid_card_ctl(struct sc_card *card, unsigned long cmd, void *ptr)
+{
+	switch (cmd) {
+	case SC_CARDCTL_GET_DEFAULT_KEY:
+		return fineid_get_default_key(card,
+				(struct sc_cardctl_default_key *) ptr);
+	case SC_CARDCTL_GET_SERIALNR:
+		return fineid_get_serialnr(card, (struct sc_serial_number *)ptr);
+	default:
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+	}
+}
+
+
+static int
+fineid_logout(struct sc_card *card)
+{
+	int rv;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	rv = iso7816_logout(card, FINEID_PIN_AUTH);
+	LOG_TEST_RET(card->ctx, rv, "AUTH PIN logout failed");
+
+	rv = iso7816_logout(card, FINEID_PIN_SIGN);
+	LOG_TEST_RET(card->ctx, rv, "SIGN PIN logout failed");
+
+	rv = iso7816_logout(card, FINEID_PIN_PUK);
+	LOG_TEST_RET(card->ctx, rv, "PUK PIN logout failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+static int
+fineid_get_serialnr(struct sc_card *card, struct sc_serial_number *serial)
+{
+	if (!serial)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+
+	if (card->serialnr.len==0)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	memcpy(serial, &card->serialnr, sizeof(*serial));
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+static int
+fineid_check_sw(struct sc_card *card, unsigned int sw1, unsigned int sw2)
+{
+	return iso_ops->check_sw(card, sw1, sw2);
+}
```

Comment:
No need to reimplement this with default ISO operation. Just remove the `fineid_ops.check_sw = fineid_check_sw;` too.